File: node/src/resources/networks.ts
import type { PaydClient } from "../client";
import type {
  Network,
  NetworkWithHelpers,
  NetworkDiscoveryResponse,
  NetworkPaymentParams,
  RawNetwork,
  RawNetworkDiscoveryResponse,
} from "../types/networks";
import type { AccountType, TransactionType } from "../types/common";
import { PaydValidationError } from "../errors";
import { validateRequired, validateEnum } from "../validators/params";

/**
 * Network Discovery operations — discover available payment networks per country.
 */
export class Networks {
  constructor(private readonly client: PaydClient) {}

  /**
   * Discover available payment networks for a specific country.
   *
   * Call this before making Pan-African collections or payouts to get the
   * required `networkCode` and `channelId`.
   *
   * @param transactionType - "receipt" for collections, "withdrawal" for payouts
   * @param dialCode - Country dial code with + prefix (e.g., "+234" for Nigeria)
   *
   * @example
   * ```typescript
   * // Discover collection networks for Nigeria
   * const networks = await payd.networks.discover("receipt", "+234");
   *
   * // Find a specific mobile money network
   * const mtn = networks.findMobile("MTN");
   *
   * // Use its params in a payment request
   * await payd.collections.panAfrican({
   *   ...mtn.toPaymentParams(),
   *   username: "my_user",
   *   amount: 1800,
   *   phoneNumber: "+2340000000000",
   *   accountNumber: "+2340000000000",
   *   accountName: "phone",
   *   narration: "Order payment",
   *   currency: "NGN",
   *   callbackUrl: "https://example.com/webhook",
   * });
   * ```
   */
  async discover(
    transactionType: TransactionType,
    dialCode: string,
  ): Promise<NetworkDiscoveryResponse> {
    validateRequired({ transactionType, dialCode }, ["transactionType", "dialCode"]);
    validateEnum(transactionType, ["receipt", "withdrawal"], "transactionType");

    if (!dialCode.startsWith("+")) {
      throw new PaydValidationError(
        `dialCode must start with + (e.g., "+234"). Got: "${dialCode}"`,
        "dialCode",
      );
    }

    const data = (await this.client.request({
      method: "GET",
      path: "/v2/networks/grouped",
      query: {
        transaction_type: transactionType,
        dial_code: dialCode,
      },
    })) as unknown as RawNetworkDiscoveryResponse;

    return buildNetworkDiscoveryResponse(data);
  }
}

/**
 * Transform raw API response into a rich NetworkDiscoveryResponse with helpers.
 */
function buildNetworkDiscoveryResponse(
  raw: RawNetworkDiscoveryResponse,
): NetworkDiscoveryResponse {
  const defaults = (raw.defaults || []).map(transformNetwork);
  const mobile = (raw.mobile || []).map(enrichNetwork);
  const banks = (raw.banks || []).map(enrichNetwork);

  return {
    defaults,
    mobile,
    banks,

    findMobile(name: string): NetworkWithHelpers {
      const lower = name.toLowerCase();
      const found = mobile.find((n) => n.name.toLowerCase().includes(lower));
      if (!found) {
        const available = mobile.map((n) => n.name).join(", ");
        throw new PaydValidationError(
          `No mobile network found matching "${name}". Available: ${available || "none"}`,
        );
      }
      return found;
    },

    findBank(name: string): NetworkWithHelpers {
      const lower = name.toLowerCase();
      const found = banks.find((n) => n.name.toLowerCase().includes(lower));
      if (!found) {
        const available = banks.map((n) => n.name).join(", ");
        throw new PaydValidationError(
          `No bank network found matching "${name}". Available: ${available || "none"}`,
        );
      }
      return found;
    },
  };
}

/** Transform a raw snake_case network into a camelCase Network. */
function transformNetwork(raw: RawNetwork): Network {
  return {
    id: raw.id,
    code: raw.code,
    updatedAt: raw.updated_at,
    status: raw.status,
    accountNumberType: raw.account_number_type as AccountType,
    country: raw.country,
    name: raw.name,
    channelIds: raw.channel_ids || [],
    selectedChannelId: raw.selected_channel_id,
    countryAccountNumberType: raw.country_account_number_type,
    minAmount: raw.min_amount,
    maxAmount: raw.max_amount,
  };
}

/** Transform and enrich a network with helper methods. */
function enrichNetwork(raw: RawNetwork): NetworkWithHelpers {
  const network = transformNetwork(raw);

  return {
    ...network,
    toPaymentParams(): NetworkPaymentParams {
      return {
        networkCode: network.id,
        channelId: network.selectedChannelId,
        transactionChannel: network.accountNumberType,
        providerName: network.name,
        providerCode: network.code,
      };
    },
  };
}

File: python/src/payd/resources/networks.py
"""Network discovery for Pan-African payments."""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING, Any

from payd.errors import PaydValidationError
from payd.validators import validate_enum, validate_required

if TYPE_CHECKING:
    from payd.client import PaydClient


@dataclass(frozen=True)
class DiscoveredNetwork:
    """Single network entry from discovery."""

    id: str
    code: str
    updated_at: str
    status: str
    account_number_type: str
    country: str
    name: str
    channel_ids: list[str]
    selected_channel_id: str
    country_account_number_type: str
    min_amount: int | float
    max_amount: int | float

    def to_payment_params(self) -> dict[str, str]:
        """Fields ready to pass into Pan-African collection or payout calls."""
        return {
            "network_code": self.id,
            "channel_id": self.selected_channel_id,
            "transaction_channel": self.account_number_type,
            "provider_name": self.name,
            "provider_code": self.code,
        }


class NetworkDiscoveryResult:
    """Rich discovery result with lookup helpers."""

    def __init__(
        self,
        *,
        defaults: list[DiscoveredNetwork],
        mobile: list[DiscoveredNetwork],
        banks: list[DiscoveredNetwork],
    ) -> None:
        self.defaults = defaults
        self.mobile = mobile
        self.banks = banks

    def find_mobile(self, name: str) -> DiscoveredNetwork:
        lower = name.lower()
        found = next((n for n in self.mobile if lower in n.name.lower()), None)
        if not found:
            available = ", ".join(n.name for n in self.mobile)
            raise PaydValidationError(
                f'No mobile network found matching "{name}". Available: {available or "none"}',
            )
        return found

    def find_bank(self, name: str) -> DiscoveredNetwork:
        lower = name.lower()
        found = next((n for n in self.banks if lower in n.name.lower()), None)
        if not found:
            available = ", ".join(n.name for n in self.banks)
            raise PaydValidationError(
                f'No bank network found matching "{name}". Available: {available or "none"}',
            )
        return found


def _parse_network(raw: dict[str, Any]) -> DiscoveredNetwork:
    ch_ids = raw.get("channel_ids") or []
    if not isinstance(ch_ids, list):
        ch_ids = []
    return DiscoveredNetwork(
        id=str(raw.get("id") or ""),
        code=str(raw.get("code") or ""),
        updated_at=str(raw.get("updated_at") or ""),
        status=str(raw.get("status") or ""),
        account_number_type=str(raw.get("account_number_type") or ""),
        country=str(raw.get("country") or ""),
        name=str(raw.get("name") or ""),
        channel_ids=[str(x) for x in ch_ids],
        selected_channel_id=str(raw.get("selected_channel_id") or ""),
        country_account_number_type=str(raw.get("country_account_number_type") or ""),
        min_amount=raw.get("min_amount") if raw.get("min_amount") is not None else 0,
        max_amount=raw.get("max_amount") if raw.get("max_amount") is not None else 0,
    )


def _build_discovery(raw: dict[str, Any]) -> NetworkDiscoveryResult:
    defaults_raw = raw.get("defaults") or []
    mobile_raw = raw.get("mobile") or []
    banks_raw = raw.get("banks") or []
    if not isinstance(defaults_raw, list):
        defaults_raw = []
    if not isinstance(mobile_raw, list):
        mobile_raw = []
    if not isinstance(banks_raw, list):
        banks_raw = []

    defaults = [_parse_network(x) for x in defaults_raw if isinstance(x, dict)]
    mobile = [_parse_network(x) for x in mobile_raw if isinstance(x, dict)]
    banks = [_parse_network(x) for x in banks_raw if isinstance(x, dict)]

    return NetworkDiscoveryResult(defaults=defaults, mobile=mobile, banks=banks)


class Networks:
    """Network discovery operations."""

    def __init__(self, client: PaydClient) -> None:
        self._client = client

    def discover(self, transaction_type: str, dial_code: str) -> NetworkDiscoveryResult:
        validate_required({"transaction_type": transaction_type, "dial_code": dial_code}, ["transaction_type", "dial_code"])
        validate_enum(transaction_type, ["receipt", "withdrawal"], "transaction_type")
        if not dial_code.startswith("+"):
            raise PaydValidationError(
                f'dial_code must start with + (e.g., "+234"). Got: "{dial_code}"',
                field="dial_code",
            )

        data = self._client.request(
            method="GET",
            path="/v2/networks/grouped",
            query={
                "transaction_type": transaction_type,
                "dial_code": dial_code,
            },
        )
        return _build_discovery(data)
